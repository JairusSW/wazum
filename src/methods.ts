import { Variant, VariantExpr, VariantInstr, VariantStmt } from './variants';

export type DataType = 'i32' | 'i64' | 'f32' | 'f64' | 'none';

const methods = {
  local: <T extends DataType>(dataType: T) => ({
    get: (name: string): Variant.LocalGet<T> => ({
      __nodeType: 'localGet',
      dataType,
      name,
      returnType: dataType,
    }),
    set: (name: string, value: VariantExpr): Variant.LocalSet => ({
      __nodeType: 'localSet',
      dataType,
      name,
      value,
      returnType: 'none',
    }),
  }),
  add:
    <T extends DataType>(dataType: T) =>
    (left: VariantExpr<T>, right: VariantExpr<T>): Variant.Add<T> => ({
      __nodeType: 'add',
      dataType,
      left,
      right,
      returnType: dataType,
    }),
  // func:
  //   <T extends DataType>(dataType: T) =>
  //   (
  //     name: string,
  //     params: [type: DataType, name: string][],
  //     locals: [type: DataType, name: string][],
  //     body: VariantExpr,
  //   ): Variant.Func<T> => ({
  //     __nodeType: 'func',
  //     name,
  //     params,
  //     locals,
  //     body,
  //     dataType,
  //   }),
  const:
    <T extends DataType>(dataType: T) =>
    (value: number): Variant.Const<T> => ({
      __nodeType: 'const',
      value,
      dataType,
      returnType: dataType,
    }),
};

export const i32 = {
  local: methods.local('i32'),
  add: methods.add('i32'),
  const: methods.const('i32'),
  // block: methods.block('i32'),
  // func: methods.func('i32'),
};
export const i64 = {
  local: methods.local('i64'),
  add: methods.add('i64'),
  const: methods.const('i64'),
  // block: methods.block('i64'),
  // func: methods.func('i64'),
};
export const f32 = {
  local: methods.local('f32'),
  add: methods.add('f64'),
  const: methods.const('f64'),
}
export const none = {
  // block: methods.block('none'),
  // func: methods.func('none'),
};

export const func = <T extends DataType, R extends T>(
  name: string,
  signature: {
    params: [type: DataType, name: string][];
    locals: [type: DataType, name: string][];
    returnType: T;
  },
  body: VariantExpr<R>,
): Variant.Func<T> => ({
  __nodeType: 'func',
  name,
  params: signature.params,
  locals: signature.locals,
  body,
  dataType: signature.returnType,
  exportName: null,
});

export const drop = (value: VariantExpr): Variant.Drop => ({
  __nodeType: 'drop',
  value,
  returnType: 'none',
});

export const block = <T extends DataType, R extends T>(
  name: string | null,
  returnType: T,
  value: [
    ...VariantStmt[],
    Omit<VariantInstr, 'returnType'> & { returnType: R },
  ],
): Variant.Block<T> => ({
  __nodeType: 'block',
  name,
  value,
  returnType,
});

export const call = <T extends DataType>(
  name: string,
  signature: {
    returnType: T;
  },
  args: VariantExpr[],
): Variant.Call<T> => ({
  __nodeType: 'call',
  name,
  args,
  returnType: signature.returnType,
});
// export const block = <Ret extends DataType>
// type VarsIntoObject<Vars extends NodeLocal[]> = {[Prop in keyof Vars[number]]: }

// type Test = VarsIntoObject<[{__nodeType: 'local', name: 'a'}]>

// export class Func {
//   /**
//    * Creates a function without adding it to a module.
//    *
//    * @param opts.name The name of the function
//    */
//   constructor(name: string, params: NodeLocal[], body: () => any[]) {}
// }

// const foo = func('foo', [i32.local('a')], [i32.local('b')], () => {
//   return [
//     b.set((i32.local('a').get(), i32.const(7))),

//     i32.add(b.get(), i32.const(42)),
//   ];
// });

// const m = new Module();
// m.addFunc(foo);
