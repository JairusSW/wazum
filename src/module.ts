import { DataType } from './methods';
import { Variant } from './variants';

export class Module {
  private funcs: Variant.Func<DataType>[] = [];

  constructor() {}

  addFunc = (func: Variant.Func<DataType>) => {
    this.funcs.push(func);
  };

  
}
