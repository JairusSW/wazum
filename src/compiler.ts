import { Module } from './module';
import { Variant, VariantInstr } from './variants';
import { match } from 'ts-pattern';
import { DataType } from './methods';

const localGet = (node: Variant.LocalGet<DataType>) => {
  return `
    (${node.dataType}.get ${node.name})
  `;
};

const add = (node: Variant.Add<DataType>) => {
  return `
    (${node.dataType}.add ${instr(node.left)} ${instr(node.right)})
  `;
};

const instr = (node: VariantInstr): string => {
  return match(node)
    .with({ __nodeType: 'add' }, add)
    .with({ __nodeType: 'localGet' }, localGet)
    .otherwise(() => {
      throw new Error(`Unexpected ${node.__nodeType} node`);
    });
};

export const compile = (m: Module) => {
  return `
    (module
      ${m.funcs
        .filter((func) => func.exportName)
        .map(
          (func) => `
        (export "${func.exportName}" (func $${func.name})
      `,
        )
        .join('\n')}

      ${m.funcs
        .map(
          (func) => `
        (func $${func.name}
          ${func.params
            .map(([dt, name]) => `(param $${name} ${dt})`)
            .join('\n')}
          (result ${func.dataType})
          ${func.locals
            .map(([dt, name]) => `(local $${name} ${dt})`)
            .join('\n')}
          ${instr(func.body)}
        )
      `,
        )
        .join('\n')}
    )
  `;
};
