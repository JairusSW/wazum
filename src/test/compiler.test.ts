import { compile } from '../compiler';
import { func, i32 } from '../methods';
import { Module } from '../module';

describe('compiler', () => {
  const addFunc = () =>
    func(
      'add',
      {
        params: [
          ['i32', 'a'],
          ['i32', 'b'],
        ],
        locals: [],
        returnType: 'i32',
      },
      i32.add(i32.local.get('a'), i32.local.get('b')),
    );

  test('add function (local.get, i32.add, exported func)', () => {
    const m = new Module();
    m.addFunc(addFunc(), true);

    expect(compile(m)).toMatchSnapshot();
  });
});
