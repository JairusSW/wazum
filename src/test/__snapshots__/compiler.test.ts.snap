// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compiler add function (local.get, i32.add, exported func) 1`] = `
"
    (module
      
        (export "add" (func $add)
      

      
        (func $add
          (param $a i32)
(param $b i32)
          (result i32)
          
          
    (i32.add 
    (i32.get a)
   
    (i32.get b)
  )
  
        )
      
    )
  "
`;
